lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -0,0 +1,485 @@
+
+// REVIEW (doilij): Where are these defined that is safe to include in PlatformAgnostic?
+typedef int16_t int16;
+typedef int32_t int32;
+typedef int64_t int64;
+typedef uint16_t uint16;
+typedef uint32_t uint32;
+typedef uint64_t uint64;
+
+#include "Codex/Utf8Codex.h"
+
+#ifndef _WIN32
+// REVIEW (doilij): The PCH allegedly defines enough stuff to get AssertMsg to work -- why was compile failing for this file?
+#ifdef AssertMsg
+#undef AssertMsg
+#endif
+#define AssertMsg(test, message)
+#ifdef AssertOrFailFastMsg
+#undef AssertOrFailFastMsg
+#endif
+#define AssertOrFailFastMsg(test, message)
+#endif // !_WIN32
+
+// REVIEW (doilij): copied from lib/Common/CommonPal.h because including that file broke the compile on Linux.
+// Only VC compiler support overflow guard
+#if defined(__GNUC__) || defined(__clang__)
+#define DECLSPEC_GUARD_OVERFLOW
+#else // Windows
+#define DECLSPEC_GUARD_OVERFLOW __declspec(guard(overflow))
+#endif
+
+#include "Common/MathUtil.h"
+#include "Core/AllocSizeMath.h"
+
+#include "Intl.h"
+#include "IPlatformAgnosticResource.h"
+
+#define U_STATIC_IMPLEMENTATION
+#define U_SHOW_CPLUSPLUS_API 1
+#pragma warning(push)
+#pragma warning(disable:4995) // deprecation warning
+#include <unicode/uloc.h>
+#include <unicode/numfmt.h>
+#include <unicode/enumset.h>
+#include <unicode/decimfmt.h>
+#pragma warning(pop)
+
+namespace PlatformAgnostic
+{
+namespace Intl
+{
+    using namespace PlatformAgnostic::Resource;
+
+    bool IsWellFormedLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch)
+    {
+        bool success = false;
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+        char icuLocaleId[ULOC_FULLNAME_CAPACITY] = { 0 };
+        char icuLangTag[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t parsedLength = 0;
+        int32_t forLangTagResultLength = 0;
+        int32_t toLangTagResultLength = 0;
+        size_t inputLangTagUtf8SizeActual = 0;
+
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        unsigned char *inputLangTagUtf8 = new unsigned char[inputLangTagUtf8SizeAllocated];
+        StringBufferAutoPtr<unsigned char> guard(inputLangTagUtf8);
+
+        if (!inputLangTagUtf8)
+        {
+            AssertOrFailFastMsg(false, "OOM: failed to allocate inputLangTagUtf8");
+            goto LReturn;
+        }
+
+        // explicitly zero the allocated array
+        for (size_t index = 0; index < inputLangTagUtf8SizeAllocated; ++index)
+        {
+            inputLangTagUtf8[index] = 0;
+        }
+
+        inputLangTagUtf8SizeActual = utf8::EncodeIntoAndNullTerminate(inputLangTagUtf8, languageTag, cch);
+
+        // Convert input language tag to a locale ID for use in uloc_toLanguageTag API.
+        forLangTagResultLength = uloc_forLanguageTag(reinterpret_cast<const char *>(inputLangTagUtf8),
```

Comment:
I'd like it if we didn't need to have this `reinterpret_cast` but I suppose `utf8::EncodeIntoAndNullTerminate` takes an `unsigned char*`, right? Can you leave out the `const` inside the cast, at least?

Reply:
All of the UTF8 stuff seems to work in terms of utf8char_t, which is typedef'd from BYTE, which is typedef'd from unsigned char. I don't know what our default warnings are but I think that could cause some (I know signed/unsigned comparison is a gcc/clang warning that could come up)

Reply:
I think this cast is necessary to some degree because of the way the types are set up in existing code, yeah. Not sure I see why leaving off the `const` is an improvement?

Reply:
Shorter line length? That's why I have been leaving them off.

Reply:
Ah, yeah. I'm adding a `const` -- no need for that, the function call does it. I was worried in general about dropping a const, although IIRC I think casting away `const` is the only thing `reinterpret_cast` won't do compared to a C-style cast, and that's a good thing :)

Anyway I'll drop `const` here and in similar places.